### ISSUE TITLE
Add a Reweighted Cross-Entropy Loss for Improved Handling of Zipfian Distributions

### ISSUE THREAD

**Issue by gkielian** [ReaLLMASIC/nanoGPT#357]
Extending observation of the following paper from NeurIPS 2024 into the logits:
https://arxiv.org/abs/2411.00680

We can easily (and per language) determine IPA frequency and subword frequency, and adjust the rewards to inhibit more frequent tokens.

Afterwards we should test to see if we require lower temperature (if the result is more robust against repeating words like "the the the" and the like.

For consideration:
```
import torch
import torch.nn as nn
import torch.nn.functional as F

class ReweightedCrossEntropyLoss(nn.Module):
    def __init__(self, frequencies, alpha=1.0):
        """
        Args:
            frequencies: A tensor of token frequencies (shape: vocab_size).
            alpha: A hyperparameter to control the strength of reweighting.
                   Higher alpha means stronger emphasis on rare tokens.
        """
        super().__init__()
        weights = 1.0 / (frequencies ** alpha)
        self.weights = weights / weights.sum()  # Normalize to sum to 1

    def forward(self, logits, targets):
        """
        Args:
            logits: Model's output logits (shape: batch_size, seq_len, vocab_size).
            targets: Ground truth token indices (shape: batch_size, seq_len).
        """
        # Calculate standard cross-entropy loss
        ce_loss = F.cross_entropy(logits.view(-1, logits.size(-1)), targets.view(-1), reduction='none')

        # Get the weights corresponding to the target tokens
        target_weights = self.weights[targets.view(-1)]

        # Reweight the loss
        weighted_loss = ce_loss * target_weights

        # Return the mean loss
        return weighted_loss.mean()

# Example Usage
frequencies = torch.tensor([0.05, 0.001, 0.03, 0.02, 0.005]) # Example frequencies
vocab_size = len(frequencies)

loss_fn = ReweightedCrossEntropyLoss(frequencies, alpha=0.5) # Experiment with alpha

# During training:
# logits = model(input_ids)
# loss = loss_fn(logits, target_ids)
# loss.backward()
# ...
```
